File: PyEngineering/08_config_loader/08_config_loader_explanation.py
"""
08 — Config Loader
====================

WHAT
----
Build a layered configuration loader for a small service:

    defaults (in code)  →  TOML file (optional)  →  environment variables

Each layer overrides the one before it. The result is validated into a
frozen `AppConfig` dataclass with type coercion and range/format checks, so
the rest of the codebase gets a fully-typed, already-validated config object
instead of passing raw dicts around.

WHY THIS MATTERS
----------------
Real services need config from multiple sources: a checked-in default, an
environment-specific file (dev/staging/prod), and per-deploy environment
variable overrides (the twelve-factor-app pattern — secrets and per-instance
values come from env vars, not files, so they never land in source control
or a shared config file). Getting the *layering order* and *validation
timing* right matters: validation must happen once, after all layers are
merged, so a bad override anywhere fails loudly at startup — not deep inside
request handling three hours later.

This module intentionally uses **only** `tomllib` (stdlib, read-only, 3.11+)
and `dataclasses` — no `pydantic` — per this curriculum's fixed toolchain
choice for problem 08: dataclasses + `__post_init__` cover real-world
validation needs without a dependency, and force you to understand what a
validation library does for you under the hood.

LAYERING SPEC
-------------
1. **Defaults** — hardcoded literal values in `DEFAULTS: dict[str, Any]`
   below (host, port, timeout, log level, feature flags).
2. **TOML file** — if a path is given and exists, `tomllib.load()` it and
   shallow-merge its top-level keys over the defaults. A missing file is not
   an error (config files are often optional/environment-specific); a
   malformed one (`tomllib.TOMLDecodeError`) must raise immediately.
3. **Environment variables** — for each config field, an env var named
   `APP_<FIELD_NAME_UPPER>` (e.g. `APP_PORT`, `APP_LOG_LEVEL`) overrides
   whatever the previous layers produced. Env vars are always strings, so
   this layer must coerce them to the field's target type (`int`, `bool`,
   etc.) — a bad coercion (e.g. `APP_PORT=not-a-number`) must raise a clear
   `ConfigError`, not an opaque `ValueError` from deep inside `int()`.
4. **Validation** — construct `AppConfig(**merged)`; `__post_init__` checks
   ranges/formats (port in 1-65535, log level in a known set, timeout > 0)
   and raises `ConfigError` with a message naming the offending field.

SPEC
----
    class ConfigError(Exception): ...

    @dataclass(frozen=True)
    class AppConfig:
        host: str = "0.0.0.0"
        port: int = 8080
        timeout_seconds: float = 30.0
        log_level: str = "INFO"
        debug: bool = False

        def __post_init__(self) -> None: ...  # validation

    DEFAULTS: dict[str, Any]  # matches AppConfig field defaults

    def load_config(
        toml_path: Path | None = None,
        *,
        env: Mapping[str, str] | None = None,   # defaults to os.environ
    ) -> AppConfig:
        "Merge DEFAULTS -> toml_path (if present) -> env vars, then
        construct/validate an AppConfig."

ACCEPTANCE CRITERIA
--------------------
- Calling `load_config()` with no arguments and no relevant env vars set
  returns an `AppConfig` equal to one built from `DEFAULTS` alone.
- A TOML file overriding a subset of keys leaves the rest at defaults.
- An env var overrides both the default AND a TOML-file value for the same
  key (env wins — it's the last layer).
- `APP_PORT="abc"` raises `ConfigError` (not a raw `ValueError`/`TypeError`).
- `port=70000` (out of range, however it got there — file or env) raises
  `ConfigError` from `__post_init__`.
- `log_level="TRACE"` (not in the allowed set) raises `ConfigError`.
- A missing TOML path is silently ignored (defaults/env still apply); a
  TOML file with invalid syntax raises (does not silently fall back).
- `AppConfig` is frozen/immutable (accidental mutation after startup is a
  bug class this should make impossible) — attempting to set an attribute
  after construction raises `dataclasses.FrozenInstanceError`.
- The `env` parameter (injectable mapping) means tests never need to
  actually mutate `os.environ` — this is the config-loader analogue of
  problem 07's "always take root as a parameter" testability lesson.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

DEFAULTS: dict[str, Any] = {
    "host": "0.0.0.0",
    "port": 8080,
    "timeout_seconds": 30.0,
    "log_level": "INFO",
    "debug": False,
}

_VALID_LOG_LEVELS = frozenset({"DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"})


class ConfigError(Exception):
    """Raised for any config loading, coercion, or validation failure."""


@dataclass(frozen=True)
class AppConfig:
    host: str = "0.0.0.0"
    port: int = 8080
    timeout_seconds: float = 30.0
    log_level: str = "INFO"
    debug: bool = False

    def __post_init__(self) -> None:
        """TODO: validate port range, timeout > 0, log_level membership.

        Raise ConfigError (not assert / bare ValueError) naming the field
        and the bad value.
        """
        raise NotImplementedError


def _load_toml_layer(toml_path: Path | None) -> dict[str, Any]:
    """TODO: return {} if toml_path is None or doesn't exist. Otherwise
    tomllib.load() it (binary mode!) and return its top-level dict. Let
    tomllib.TOMLDecodeError propagate (or wrap in ConfigError — pick one and
    be consistent)."""
    raise NotImplementedError


def _coerce_env_value(field_name: str, raw: str, target_type: type) -> Any:
    """TODO: coerce the string env var to target_type.

    - bool: accept a small known set of truthy/falsy strings case-
      insensitively (e.g. "true"/"false", "1"/"0") — do NOT use `bool(raw)`,
      which is True for any non-empty string including "false"!
    - int/float: wrap the conversion, raise ConfigError with field name on
      failure instead of letting ValueError leak out raw.
    - str: pass through unchanged.
    """
    raise NotImplementedError


def _load_env_layer(env: Mapping[str, str]) -> dict[str, Any]:
    """TODO: for each field in AppConfig's dataclass fields, look up
    APP_<FIELD_UPPER> in env; if present, coerce via _coerce_env_value using
    the field's declared type and add to the returned overrides dict.
    """
    raise NotImplementedError


def load_config(
    toml_path: Path | None = None,
    *,
    env: Mapping[str, str] | None = None,
) -> AppConfig:
    """TODO: merge DEFAULTS -> toml layer -> env layer (env parameter
    defaults to os.environ when None), then construct AppConfig(**merged).
    Let AppConfig.__post_init__'s ConfigError propagate.
    """
    raise NotImplementedError


# ---------------------------------------------------------------------------
# Hints
# ---------------------------------------------------------------------------
# - `tomllib.load(f)` requires `f` opened in binary mode ("rb") — it raises
#   TypeError on a text-mode file, unlike `json.load`.
# - `dataclasses.fields(AppConfig)` gives you (name, type) pairs to drive
#   the env-var layer generically instead of hand-writing one branch per
#   field.
# - `field.type` on a dataclass field can be a string (from `from __future__
#   import annotations`) rather than the real type object — if you hit this,
#   either drive coercion off `DEFAULTS`'s value types with `type(default)`,
#   or use `typing.get_type_hints(AppConfig)` to resolve string annotations.
# - Keep `_coerce_env_value` and `__post_init__` doing different jobs:
#   coercion turns strings into the right Python type; validation checks
#   that the (already-typed) value is in range/allowed.
#
# Pitfalls
# --------
# - `bool("false")` is `True` — a classic env-var config bug. Never use the
#   `bool` constructor directly on a raw env string.
# - Deep-merging when you meant shallow-merging (or vice versa) — this spec
#   is intentionally shallow (flat config), so `dict.update` semantics are
#   sufficient; note in your solution why a nested config would need real
#   recursive merging.
# - Validating *before* all layers are merged — a value that's invalid at
#   the defaults layer but corrected by an env var should not raise.
# - Swallowing `tomllib.TOMLDecodeError` silently — a malformed file should
#   be loud, not treated the same as a missing one.
#
# Stretch goals
# -------------
# - Support nested TOML tables mapped to nested dataclasses.
# - Add a `--dump-effective-config` style `to_dict()` / logging helper that
#   redacts fields named like secrets before printing (real services must
#   never log raw secrets on startup).
# - Add a `source: dict[str, str]` companion result recording which layer
#   won for each field (defaults/file/env) — invaluable for debugging "why
#   is prod using the wrong port" incidents.

File: GoEngineering/08_config_loader/explanation/08_config_loader_explanation.go
/*
Package config — Layered Config Loader (Problem 08)

WHAT WE'RE BUILDING

A config loader for a hypothetical service, with three layers applied in
increasing priority:

 1. Defaults embedded into the binary at compile time (embed.FS).
 2. A JSON config file on disk, if present, overriding the defaults.
 3. Environment variables, overriding both, for per-deployment secrets and
    ops-controlled tuning (timeouts, log level) without a redeploy.

Then everything is validated once, at the end, so callers get ONE clear error
listing everything wrong instead of failing on the first bad field, restarting,
failing on the second, and so on — a real operational annoyance in services
with many required settings.

# WHY IT MATTERS IN REAL SYSTEMS

Every non-trivial service needs configuration from more than one source: a
baseline that ships with the binary (so it runs with sane defaults out of the
box), an optional file for environment-specific overrides (staging vs prod
config, checked into a separate ops repo), and environment variables for the
handful of values that must never live in a file (API keys, DB passwords) or
that ops wants to flip without touching a file (LOG_LEVEL=debug for one
incident). Getting the *precedence* right and *validating once with a
complete error report* are the two things that separate a config loader that
survives a real on-call rotation from one that causes a 2am guessing game.

CONCEPTS COVERED
  - embed.FS for compiling a defaults file directly into the binary (no
    "make sure defaults.json ships next to the binary" deployment footgun)
  - Layered override: struct-level merge where a "present" field in a later
    layer wins, but an *absent* field in a later layer keeps the earlier
    layer's value — this needs pointer fields or an explicit "was this set"
    tracking mechanism, not plain zero-value structs (a zero-value bool
    false is indistinguishable from "not set" otherwise)
  - Environment variable parsing with type conversion and per-field error
    accumulation (strconv errors reported per-field, not the first one
    aborting the whole load)
  - Validation as a single pass returning errors.Join of every violation,
    not fail-fast on the first bad field
  - time.Duration parsing from both JSON (string form, "30s") and env
    (string form as well, but a different unmarshal path)

REQUIREMENTS / SPEC

	//go:embed defaults.json
	var defaultsFS embed.FS
	    The compiled-in baseline. Ships inside the binary; see the solution
	    file for the actual embed directive and JSON content.

	type Config struct {
	    ServerAddr   string        `json:"server_addr" env:"SERVER_ADDR"`
	    ReadTimeout  time.Duration `json:"read_timeout" env:"READ_TIMEOUT"`
	    WriteTimeout time.Duration `json:"write_timeout" env:"WRITE_TIMEOUT"`
	    MaxConns     int           `json:"max_conns" env:"MAX_CONNS"`
	    LogLevel     string        `json:"log_level" env:"LOG_LEVEL"`
	    DatabaseDSN  string        `json:"database_dsn" env:"DATABASE_DSN"`
	}
	    The fully-resolved config a service actually runs with. Durations are
	    JSON strings ("30s") that unmarshal via time.Duration's
	    UnmarshalJSON-compatible representation — actually time.Duration does
	    NOT implement json.Unmarshaler by default (it (un)marshals as a plain
	    int64 nanoseconds count), which is a very common gotcha; the solution
	    must handle "30s"-style strings explicitly.

	type Loader struct { ... }
	    Holds the embedded defaults FS and an environ-reading function (make
	    it injectable — func() []string — instead of calling os.Environ()
	    directly, so tests can supply a fixed environment without mutating
	    real process env vars, which is inherently racy across parallel
	    tests).

	func NewLoader(defaults embed.FS) *Loader
	func (l *Loader) Load(filePath string) (Config, error)
	    1. Parse the embedded defaults.json into a Config (always succeeds —
	       it's compiled in and should be a build-breaking test failure if
	       it's ever invalid JSON).
	    2. If filePath != "" and the file exists, parse it and overlay onto
	       the defaults (fields present in the file's JSON override; fields
	       absent keep the default). If filePath != "" and the file does NOT
	       exist, treat that as an error (a caller who names a file expects
	       it to be there) — but filePath == "" is a valid "no override file"
	       signal, not an error.
	    3. Overlay environment variables (using the injected environ
	       function): for each struct field with an `env:"NAME"` tag, if
	       NAME is set in the environment, parse it into the field's type
	       and overlay it. Collect (don't abort on) individual parse errors
	       (e.g. MAX_CONNS=notanumber) and return them all joined via
	       errors.Join at the end alongside any validation errors.
	    4. Validate the final merged Config (see Validate below). Return the
	       Config and a joined error containing every problem found across
	       env-parsing and validation, or (Config, nil) if everything's
	       clean.

	func (c Config) Validate() error
	    Returns errors.Join of every violation (not just the first):
	      - ServerAddr must be non-empty.
	      - ReadTimeout and WriteTimeout must be > 0.
	      - MaxConns must be > 0.
	      - LogLevel must be one of "debug", "info", "warn", "error".
	      - DatabaseDSN must be non-empty.

ACCEPTANCE CRITERIA
  - Precedence is exactly defaults < file < env, verified by a test that sets
    all three layers to different values for the same field and asserts env
    wins.
  - A field absent from the override file does NOT get zeroed out — it keeps
    the default (this is the part naive `json.Unmarshal(data, &cfg)` onto an
    already-populated struct actually gets right for missing keys, but ONLY
    if you unmarshal onto the existing Config rather than a fresh zero-value
    one — the solution must explain this explicitly).
  - filePath == "" loads cleanly from defaults+env only.
  - filePath naming a nonexistent file is an error.
  - Multiple validation failures at once are all present in the returned
    error (use errors.Join + errors.Is checks against sentinel errors in the
    test, or substring checks on the joined message).
  - Malformed env values (e.g. MAX_CONNS=abc) produce a clear, field-named
    error rather than a panic or a silently-ignored bad value.

Read the HINTS and PITFALLS at the bottom before opening the solution file.
*/
package config

import (
	"embed"
	"errors"
	"time"
)

// Config is the fully-resolved configuration a service runs with.
type Config struct {
	ServerAddr   string        `json:"server_addr" env:"SERVER_ADDR"`
	ReadTimeout  time.Duration `json:"read_timeout" env:"READ_TIMEOUT"`
	WriteTimeout time.Duration `json:"write_timeout" env:"WRITE_TIMEOUT"`
	MaxConns     int           `json:"max_conns" env:"MAX_CONNS"`
	LogLevel     string        `json:"log_level" env:"LOG_LEVEL"`
	DatabaseDSN  string        `json:"database_dsn" env:"DATABASE_DSN"`
}

// Loader resolves a Config from embedded defaults, an optional override
// file, and environment variables, in that ascending priority order.
type Loader struct {
	// TODO: hold the embedded defaults FS and an "environ" func() []string
	// (default os.Environ, but overridable for hermetic tests).
}

// NewLoader returns a Loader that reads its baseline defaults from
// defaults (expected to contain a "defaults.json" entry).
func NewLoader(defaults embed.FS) *Loader {
	// TODO: store defaults and default the environ func to os.Environ.
	panic("TODO: implement NewLoader")
}

// Load resolves the final Config: embedded defaults, overlaid by filePath's
// JSON if filePath != "", overlaid by environment variables tagged `env:"…"`
// on Config's fields. filePath == "" skips the file layer. Returns every
// env-parsing and validation problem joined together, or a valid Config and
// nil.
func (l *Loader) Load(filePath string) (Config, error) {
	// TODO:
	//  1. Read+unmarshal defaults.json from the embedded FS into a Config.
	//     This should never fail in a correctly-built binary; if it does,
	//     that's a programmer error worth a distinct wrapped message.
	//  2. If filePath != "": os.ReadFile it (missing file => error, do not
	//     silently ignore); json.Unmarshal ONTO the existing defaults
	//     Config value (not a fresh zero-value one!) so JSON keys absent
	//     from the file leave the corresponding fields untouched.
	//  3. Walk the Config's struct fields via reflection (or a manual
	//     switch, simpler and arguably clearer for six fields — your call,
	//     state which and why in the solution) looking for each field's
	//     `env:"NAME"` tag; if NAME is present in l.environ(), parse it
	//     into the field's type (special-case time.Duration via
	//     time.ParseDuration, everything else via strconv). Accumulate
	//     parse errors with errors.Join instead of returning on the first.
	//  4. Call cfg.Validate() and join any validation errors onto the
	//     env-parsing errors.
	//  5. Return (cfg, joinedErrOrNil).
	panic("TODO: implement Load")
}

// Validate reports every problem with c at once (via errors.Join), rather
// than failing fast on the first bad field.
func (c Config) Validate() error {
	// TODO: accumulate one error per violated rule (see spec above) into a
	// []error and return errors.Join(violations...) (errors.Join returns
	// nil for an empty/all-nil slice, so no special-casing "no errors" is
	// needed).
	panic("TODO: implement Validate")
}

// Sentinel-style validation errors a caller might want to errors.Is against
// for a specific known problem (useful for e.g. distinguishing "bad log
// level" from "missing DSN" in a caller that wants to react differently).
var (
	ErrMissingServerAddr = errors.New("config: server_addr is required")
	ErrInvalidTimeout    = errors.New("config: read/write timeout must be positive")
	ErrInvalidMaxConns   = errors.New("config: max_conns must be positive")
	ErrInvalidLogLevel   = errors.New("config: log_level must be one of debug/info/warn/error")
	ErrMissingDSN        = errors.New("config: database_dsn is required")
)

/*
HINTS
  - json.Unmarshal only overwrites fields whose keys are actually present in
    the input JSON — unmarshaling `{"log_level":"debug"}` onto a Config that
    already has ServerAddr set from defaults leaves ServerAddr untouched.
    This is exactly the layering mechanism you want, but only works if you
    unmarshal onto the PREVIOUS layer's already-populated struct, never onto
    a fresh `var cfg Config`.
  - time.Duration's default JSON representation is a raw int64 (nanoseconds),
    NOT a "30s"-style string — encoding/json has no built-in support for the
    friendly string form. To accept `"read_timeout": "30s"` in the config
    file, either (a) give Config a custom UnmarshalJSON that decodes
    durations as strings via time.ParseDuration, or (b) use a shadow struct
    with `ReadTimeout string` fields for unmarshaling and convert
    afterward. Pick one and be consistent for both the file layer and the
    env layer.
  - reflect is a legitimate, idiomatic choice here for genericizing the
    env-tag walk across all fields (viper, envconfig, and most real config
    libraries do exactly this) — but for a six-field struct, a plain
    explicit switch/if-chain reading each env var by name is also completely
    reasonable and arguably more readable/debuggable. Either is acceptable;
    just document the trade-off you picked.
  - Inject the environment reader (func() []string, or even more simply
    func(name string) (string, bool)) rather than calling os.Getenv/
    os.Environ directly inside Load — tests that need different env values
    per subtest cannot safely use os.Setenv with t.Parallel() (shared
    process-global state), but an injected function is trivially fakeable
    per test with no shared state at all.

COMMON PITFALLS
  - Unmarshaling the override file onto a fresh `var cfg Config` instead of
    the defaults-populated one — silently zeroes every field the file
    doesn't mention, which is the single most common bug in hand-rolled
    layered config loaders.
  - Treating a missing override file path as "use defaults" when filePath
    was explicitly given — masks operator typos (a mistyped --config flag
    silently running on bare defaults in production is a bad night).
  - Fail-fast validation (return on the first bad field) — forces an
    operator through N redeploy-and-retry cycles to discover all N problems
    instead of seeing the complete list once.
  - Forgetting that env var values are always strings — MAX_CONNS="10"
    needs strconv.Atoi, and a malformed value must produce a field-named
    error, not silently fall back to zero or panic.

STRETCH GOALS
  - Add a `Dump() map[string]string` that reports, per field, which layer
    (default/file/env) actually supplied its final value — invaluable for
    debugging "why is this service using the wrong timeout" in production.
  - Support a fourth layer: CLI flags (highest priority), using the
    standard `flag` package.
  - Add hot-reload: a `Watch(ctx, filePath, onChange func(Config))` that
    re-loads when the file changes (fsnotify-style, or simple polling) and
    re-validates before calling onChange, never handing the caller an
    invalid Config.
  - Redact DatabaseDSN (and any future secret field) in a String()/LogValue()
    method so accidentally logging a Config never leaks credentials.
*/
